### PR TITLE
Focus-Management

### LINKED ISSUE #4992: Focus management
## Acceptance criteria
- never lose keyboard focus
- meaningful order of focussable elements
- always stay in context / resume old context

### PR DESCRIPTION
**Pull request by fschade**
## Description
to get a better background why this is needed please read #4992 beforehand.

## Related Issue

- Fixes #4992 

## Motivation and Context
This pr introduces the concept of user driven focus management.

This is helpful when a user navigates using the keyboard and the given context switches, for example the sidebar or and a modal opens which has a different tabindex and the user will never get there (more or less). 

Now it's possible to decide which item should be focused after a given action (modal open, sidebar show). 
The important part here is that those interactions need to be able to be rolled back.  (replayed in a reverse order).

- if a user closes a modal, the focus should jump back to the exact position where he started.
- if a sidebar opens the focus should jump into the sidebar or vice versa if it gets closed.

following worst case scenario is part of the idea behind it.
- user tabs through the page
- user opens the sidebar by pressing enter on the dedicated button
- the sidebar opens and the focus needs to jump to the close button of it
- the user tabs through the links in the sidebar and hits enter on "rename"
- the rename modal opens, the user tabs to update
- the modal closes and the focus must jump back to "rename" in sidebar
- the user tabs back to the sidebar close button and press enter which closes the sidebar
- the focus must jump back to the element where the user started (the three dots for example)

all this needs to be replayable in random order.

On top of this we also fixed some minor issues regarding tabindex in the top-bar. It also includes a new global lifecycle mixing which makes it possible to listen to events like mounted, replayable, ... without the requirement that those are defined by the consumed component.

## How Has This Been Tested?

- manual

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] tests
- [ ] discuss the scope of this
- [ ] changelog

### CODE REVIEW

File: packages/web-runtime/src/components/ApplicationsMenu.vue [hunk 1]
```diff
@@ -21,11 +21,17 @@
...
+            key="apps-menu-external-link"
+            :target="n.target"
+            :href="n.url"
+            tabindex="0"
+          >
             <oc-icon :name="n.iconMaterial" size="xlarge" />
             <span class="uk-display-block" v-text="$gettext(n.title)" />
           </a>
-          <router-link v-else key="apps-menu-internal-link" :to="n.path">
+          <router-link v-else key="apps-menu-internal-link" :to="n.path" tabindex="0">
```
Comment (pascalwengerter) [inline packages/web-runtime/src/components/ApplicationsMenu.vue:34]: `tabinde="0"` shouldn't strictly be necessary (tab navigation works in Chrome without it and doesn't work in Firefox/Safari either way)

**Review by pascalwengerter — COMMENTED** [review]
Just a question regarding tabindex, rest looks good & works like a charm!

File: packages/web-runtime/src/components/ApplicationsMenu.vue (hunk 1 above)
Comment (fschade) [reply packages/web-runtime/src/components/ApplicationsMenu.vue]: solved

**Comment by kulmann** [comment]
@fschade The initial focus of the right sidebar is the `close` button now. That seems to be wrong. IMO it should be on the sidebar itself, not on the close button. AFAIK a screen reader would start reading with the close button now instead of with the resource name.

**Comment by kulmann** [comment]
Clicking the file row instead of the three dots button results in resetting the focus to the top of the page after closing the right sidebar.

**Comment by fschade** [comment]
> @fschade The initial focus of the right sidebar is the `close` button now. That seems to be wrong. IMO it should be on the sidebar itself, not on the close button. AFAIK a screen reader would start reading with the close button now instead of with the resource name.

good point, fixed. Thanks

**Review by kulmann — APPROVED** [review]
nice 😍

**Pull request merged**